pkg/interface/src/views/apps/chat/components/lib/chat-window.tsx

```diff
@@ -126,68 +142,104 @@ export class ChatWindow extends Component {
     const { props } = this;
 
     if (
-      props.messages.length >= props.length ||
-      this.hasAskedForMessages ||
-      props.length <= 0
+      props.envelopes.length >= props.mailboxSize || // Can't fetch more messges than exist
+      this.fetchPending || // Can't fetch if currently fetching
+      props.mailboxSize <= 0 // Can't fetch if no messages exist
     ) {
       return;
     }
 
     const start =
-      props.length - props.messages[props.messages.length - 1].number;
+      props.mailboxSize - props.envelopes[props.envelopes.length - 1].number;
     if (start > 0) {
-      const end = start + size < props.length ? start + size : props.length;
+      const end = start + size < props.mailboxSize ? start + size : props.mailboxSize;
       props.api.chat.fetchMessages(start + 1, end, props.station);
-      this.hasAskedForMessages = true;
+      this.fetchPending = true;
     }
   }
 
-  render() {
-    const { props, state } = this;
-    const sliceLength = Math.min(
-      state.numPages * PAGE_SIZE,
-      props.messages.length + props.pendingMessages.length
-    );
-    const messages =
-      props.pendingMessages
-        .concat(props.messages)
-        .slice(0, sliceLength);
 
+  render() {
+    const { props: {
+      envelopes,
+      stationPendingMessages,
+      unreadCount,
+      unreadMsg,
+      isChatLoading,
+      isChatUnsynced,
+      api,
+      ship,
+      station,
+      association,
+      group,
+      contacts,
+      mailboxSize
+    }, state } = this;
+    
+    const messages: Envelope[] = [];
+
+    // Reindex messages based on their number.
+    // We want the virtual list to know the total
+    // size of messages and render based on that
+    envelopes
+      .concat(stationPendingMessages)
+      .forEach((message) => {
+        messages[message.number] = message;
+      });
+
+    const loadedMessageCount = Object.entries(messages).length;
+
+    // Virtuoso tries to scroll offscreen if the latest message
+    // can't reach the bottom, so just don't set the initial
+    // index if fewer than 20 messages exist in the chat
+    const initialIndex =
+      mailboxSize - 1 < INITIAL_LOAD
+      ? undefined
+      : unreadCount // otherwise if there are unread messages
+        ? mailboxSize - unreadCount - 1 // put the one right before at the top
+        : mailboxSize - 1; // Just load it at the end
+    
+    const overscan = 100;
```

Comment:
should probably be declared outside of the function scope as an all caps constant.